### PR TITLE
refactor: scheduler manager

### PR DESCRIPTION
**Pull request by gaius-qi**
## Description
- Refactor scheduler manager

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation Update (if none of the other choices apply)

## Checklist:

- [x] Code compiles correctly.
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: scheduler/core/evaluator/basic/basic_evaluator.go [hunk 1]
```diff
@@ -34,53 +34,58 @@ func NewEvaluator(cfg *config.SchedulerConfig) evaluator.Evaluator {
...
-		return true
+	parent, ok := peer.GetParent()
+	if !ok {
+		if !peer.IsDone() {
+			logger.Debugf("peer %s need adjust parent because it has not parent and status is %s", peer.ID, peer.GetStatus())
+			return true
+		}
+
+		// TODO Check whether the parent node is in the blacklist
+		if eval.IsBadNode(parent) {
```
Comment (244372610) [inline scheduler/core/evaluator/basic/basic_evaluator.go:49]: parent不存在的情况，不应该去做判断

File: scheduler/core/evaluator/basic/basic_evaluator.go [hunk 2]
```diff
@@ -34,53 +34,58 @@ func NewEvaluator(cfg *config.SchedulerConfig) evaluator.Evaluator {
...
+			logger.Debugf("peer %s need adjust parent because it current parent is bad", peer.ID)
+			return true
+		}
 	}
 
-	if peer.GetParent() != nil && peer.GetParent().IsLeave() {
-		logger.Debugf("peer %s need adjust parent because it current parent is status is leave", peer.PeerID)
+	if parent.IsLeave() {
```
Comment (244372610) [inline scheduler/core/evaluator/basic/basic_evaluator.go:55]: 这里如果parent不存在会panic

File: scheduler/core/monitor.go [hunk 3]
```diff
@@ -102,9 +105,9 @@ func (m *monitor) printDebugInfo() string {
 		if node == nil {
 			return
 		}
-		nPath := append(path, fmt.Sprintf("%s(%d)(%s)", node.PeerID, node.GetWholeTreeNode(), node.GetStatus()))
+		nPath := append(path, fmt.Sprintf("%s(%d)(%s)", node.ID, node.GetTreeLen(), node.GetStatus()))
```
Comment (244372610) [inline scheduler/core/monitor.go:108]: 替换成 GetTreeNodesCount ?

File: scheduler/core/service.go [hunk 4]
```diff
@@ -211,85 +211,81 @@ func (s *SchedulerService) GenerateTaskID(url string, meta *base.UrlMeta, peerID
...
-	peer = supervisor.NewPeer(req.PeerId, task, peerHost)
+	peer = supervisor.NewPeer(req.PeerId, task, host)
 	s.peerManager.Add(peer)
 	return peer
 }
 
 func (s *SchedulerService) GetOrCreateTask(ctx context.Context, task *supervisor.Task) *supervisor.Task {
 	span := trace.SpanFromContext(ctx)
-	synclock.Lock(task.TaskID, true)
+	synclock.Lock(task.ID, false)
```
Comment (244372610) [inline scheduler/core/service.go:245]: 这里为什么要替换成写锁，GetOrAdd方法中使用的是syncMap,是否可以直接把锁去掉

File: scheduler/rpcserver/rpcserver.go [hunk 5]
```diff
@@ -145,26 +141,26 @@ func (s *SchedulerServer) ReportPieceResult(stream scheduler.Scheduler_ReportPie
 		span.RecordError(err)
 		return err
 	}
-	conn := peer.BindNewConn(stream)
-	logger.Infof("peer %s is connected", peer.PeerID)
+	conn, _ := peer.BindNewConn(stream)
```
Comment (244372610) [inline scheduler/rpcserver/rpcserver.go:144]: 这里替换为接口返回error为什么没用到

File: scheduler/supervisor/host.go [hunk 6]
```diff
@@ -35,8 +69,8 @@ type PeerHost struct {
 	RPCPort int32
 	// DownloadPort piece downloading port for peer
 	DownloadPort int32
-	// CDN if host type is cdn
-	CDN bool
+	// IsCDN if host type is cdn
+	IsCDN bool
```
Comment (244372610) [inline scheduler/supervisor/host.go:73]: 布尔类型的变量前面不应该加 Is 前缀

File: scheduler/core/service.go (hunk 4 above)
Comment (gaius-qi) [reply scheduler/core/service.go:245]: 因为对 task 进行锁，后面是对 task 的连续写操作。

Comment (244372610) [reply scheduler/core/service.go:245]: 后面有转换成写锁的操作

File: scheduler/core/service.go [hunk 7]
```diff
@@ -211,85 +211,81 @@ func (s *SchedulerService) GenerateTaskID(url string, meta *base.UrlMeta, peerID
...
 	}
 
-	synclock.UnLock(task.TaskID, true)
 	// do trigger
-	task.UpdateLastTriggerTime(time.Now())
-
-	synclock.Lock(task.TaskID, false)
-	defer synclock.UnLock(task.TaskID, false)
+	task.LastTriggerAt.Store(time.Now())
 	span.SetAttributes(config.AttributeTaskStatus.String(task.GetStatus().String()))
```
Comment (244372610) [inline scheduler/core/service.go:261]: 为什么把写锁去掉，这里不会有并发问题吗？如果不加锁的话下面的CDN触发逻辑可能会执行多次

**Review by jim3ma — APPROVED** [review]
LGTM

**Review by 244372610 — APPROVED** [review]

**Pull request merged**